### ISSUE TITLE
[RFC] @lifecycleHooks decorator for sharing lifecycle hook logic

### ISSUE THREAD

**Issue by fkleuver** [aurelia/aurelia#1044]
# 💬 RFC

### NOTE: This proposal is deliberately minimalistic and intended as an MVP to be included in alpha

This RFC proposes a unified API to share common logic between components, intended to be run at specific lifecycle hook timings.

The `@lifecycleHooks` decorator will be a direct replacement for v1's `@viewEngineHooks`, as well as v1 router's pipeline slots, with improved consistency and leaving room for (potentially unknown) future enhancements.

Unlike `@viewEngineHooks`, there is not a pre-determined set of hooks that can be tapped into. This allows users and plugin authors to define new hooks if they want to, in much the same way that Aurelia itself (for starters the runtime and the router) would be utilizing it.

### Might address these issues and/or is related to
- https://github.com/aurelia/aurelia/issues/978
- https://github.com/aurelia/aurelia/issues/1037

### TLDR (what does the API look like?)

```js
// do-stuff.js
export class DoStuffLifecycleHooks {
  binding(vm, ...args) {}
  attaching(vm, ...args) {}
  detaching(vm, ...args) {}
  // etc
}

// consumer.js
export class ConsumerCustomElement {
  static dependencies = [DoStuff]; // Or register globally
}
```

### Use cases
A common requirement in apps is to run the same bit of logic in the same lifecycle hook in multiple components. Some examples:
1) `*`: Conditionally log each lifecycle hook invocation for advanced troubleshooting.
2) `binding`, `load`: Lazy load data from the server only once, but initiate this from whichever component (out of several that need it) happens to be rendered first.
3) `define`, `hydrating`, `hydrated`, `created`: Apply CSS to components matching arbitrary/dynamic criteria or other custom runtime conventions
4) `attaching`, `detaching`: Run the same enter/leave animations for all routed components.
5) `canLoad`: Run auth guards for all components or a (potentially dynamic) subset thereof.
6) `canUnload`: Run an "unsaved changes" dialog guard for all components that have an `isDirty` getter on them.

### Prior art
This API carries some similarities from concepts from other frameworks:
- React: [higher-order components](https://reactjs.org/docs/higher-order-components.html)
- Vue: [mixins](https://vuejs.org/v2/guide/mixins.html)
- In Angular and Aurelia 1, you would do this sort of thing either via TypeScript mixins, or manual aggregation ("composition over inheritance").

### Resource semantics

When `@lifecycleHooks` is applied to a class, that class becomes a "resource" (it's registered to the `Protocol.resource` metadata api just like `@customElement`, `@valueConverter`, etc), giving it resource-specific semantics for DI and the runtime module loader/analyzer:
- Registering it in the root DI container (e.g. via `Aurelia.register`) turns it into a global resource and causes it to be invoked for *every* component
- Adding it to the `dependencies` list of a specific component, causes it to be invoked for only that component.
- It's recognized by the module analyzer, meaning it can lazy loaded (natively or with help of webpack / other bundlers) with relative ease.
- It will be supported by conventions (by adding the `LifecycleHooks` suffix to the class name).
```ts
// explicit
@lifecycleHooks()
export class SharedStuff { ... }

// conventions:
export class SharedStuffLifecycleHooks { ... }
```

The above are pretty much freebies for each API that adheres to the resource protocol, and `@lifecycleHooks` seems like a logical consumer of this existing infrastructure.

### Contracts & behavior

- The first argument passed-in is always the ViewModel, followed by the arguments that are normally passed-in to these hooks.
- `LifecycleHook`s are *scoped singletons*, just like `ValueConverter`s and `BindingBehavior`s, meaning globally registered ones are global singletons, whereas locally registered ones are singletons only within their registered context (typically one instance per type)
- Shared hooks run *before* (and if async, in parallel with) their component instance counterparts.

### Future extensions

Depending on the feedback during alpha, we may add some quality-of-life configuration options to the `@lifecycleHooks` to enable more use cases and/or improve performance. For example:
```ts
// Control whether the hook runs before, after, or in parallel with the component instance hook.
// This would default to 'parallel' (which is the current only behavior), making this a non-breaking change.
@lifecycleHooks({ order: 'before' | 'after' | 'parallel' })
```

```ts
// Apply a compile-time filter that, when it returns true, determines whether or not this hook will run for the specified component.
// Strictly speaking this does not add a new feature, but it could counteract the potentially significant performance impact
// of many shared hooks across an app that are only meant to run on a few components
@lifecycleHooks({ match: (definition: IResourceDefinition) => boolean })
```

## 💻 Examples

Let's go through the aforementioned use cases, now with example implementations using this new API:
### 1) `*`

Conditionally log each lifecycle hook invocation for advanced troubleshooting.

- Registration scope: global

```js
// lifecycle-logger.js
@lifecycleHooks()
export class LifecycleLogger {
  constructor(@ILogger logger) { this.logger = logger; }
  define(vm) { this.trace('define', vm); }
  hydrating(vm) { this.trace('hydrating', vm); }
  hydrated(vm) { this.trace('hydrated', vm); }
  created(vm) { this.trace('created', vm); }
  binding(vm) { this.trace('binding', vm); }
  bound(vm) { this.trace('bound', vm); }
  attaching(vm) { this.trace('attaching', vm); }
  attached(vm) { this.trace('attached', vm); }
  detaching(vm) { this.trace('detaching', vm); }
  unbinding(vm) { this.trace('unbinding', vm); }
  canLoad(vm) { this.trace('canLoad', vm); return true; }
  load(vm) { this.trace('load', vm); }
  canUnload(vm) { this.trace('canUnload', vm); return true; }
  unload(vm) { this.trace('unload', vm); }
  trace(hook, vm) { this.logger.trace(`${hook} ${vm.$controller.definition.name}`); }
}

// index.js
Aurelia
  .register(
    LifecycleLogger,
    ...,
  )
  .app(...)
  .start();
```

### 2) `binding`, `load`

Lazy load data from the server only once, but initiate this from whichever component (out of several that need it) happens to be rendered first.

- Registration scope: local (added by hand to `dependencies`)

(in this example, we use it for a dropdown that needs some server-side data that only needs to be loaded once)

```js
// things-initializer.js
export const ICommonThings = DI.createInterface().withDefault(x => x.singleton(CommonThings));
export class CommonThings {
  constructor(@IHttpClient http) { this.http = http; }
  init() {
    if (!this.isInitialized) {
      return this.initPromise ??= (async () => {
        this.data = await this.http.get('api/common-things');
        this.isInitialized = true;
      })();
    }
  }
}
@lifecycleHooks
export class ThingsInitializer {
  constructor(@ICommonThings things) { this.things = things; }
  binding() { return this.things.init(); }
  load() { return this.things.init(); }
}
// some-dropdown.js
@customElement({
  name: 'some-dropdown',
  template: '<select :value="thingId"><option repeat.for="thing of things.data" :model="thing.id">${thing.name}</option></select>',
  dependencies: [ThingsInitializer],
})
export class SomeDropdown {
  constructor(@ICommonThings things) { this.things = things; }
}
```

### 3) `define`, `hydrating`, `hydrated`, `created`

Apply CSS to components matching arbitrary/dynamic criteria or other custom runtime conventions

- Registration scope: global

```js
// ripple-effect-applicator.js
@lifecycleHooks
export class RippleEffectApplicator {
  created(vm) {
    if (vm.$controller.definition.name.startsWith('bs-') && vm.someCondition) {
      // (note: this definition.instructions stuff will become easier to use soon)
      vm.$controller.definition.instructions[0].push(new HydrateAttributeInstruction('ripple', []));
    }
  }
}

// index.js
Aurelia
  .register(
    RippleEffectApplicator,
    ...,
  )
  .app(...)
  .start();
```

### 4) `attaching`, `detaching`:

Run the same enter/leave animations for all routed components.

- Registration scope: local (added by custom decorator)

```js
// animations.js
const duration = 300;
export class SlideInOutAnimations {
  attaching(vm) {
    return vm.$controller.host.animate([{ transform: 'translate(-100vw, 0)' }, { transform: 'translate(0, 0)' }], { duration, easing: 'ease-in' }).finished;
  }
  detaching(vm) {
    return vm.$controller.host.animate([{ transform: 'translate(0, 0)' }, { transform: 'translate(100vw, 0)' }], { duration, easing: 'ease-out' }).finished;
  }
}
export class FadeInOutAnimations {
  attaching(vm) {
    return vm.$controller.host.animate([{ opacity: 0 }, { opacity: 1 }], { duration }).finished;
  }
  detaching(vm) {
    return vm.$controller.host.animate([{ opacity: 1 }, { opacity: 0 }], { duration }).finished;
  }
}
// let's make a convenient decorator for ourselves this time:
export function pageTransitions(target) {
  (target.dependencies ??= []).push(SlideInOutAnimations, FadeInOutAnimations);
}

// home-page.js
@pageTransitions
export class HomePage {}

// settings-page.js
@pageTransitions
export class SettingsPage {}
```

### 5) `canLoad`:

Run auth guards for all components or a (potentially dynamic) subset thereof.

- Registration scope: global

```js
// auth-guard.js
@lifecycleHooks
export class AuthGuard {
  constructor(@IAuthService authService) { this.authService = authService; }
  // (note: router hook signatures are still subject to change)
  async canLoad(vm, params, next, current) {
    if (!current.data.auth || await this.authService.isAuthorized(current.data.auth.scopes)) {
      return true;
    }
    return `/login(reason=${current.data.auth.scopes})`;
  }
}

// settings-page.js
@route({ data: { auth: { scopes: ['settings'] } } })
export class SettingsPage {}

// profile-page.js
@route({ data: { auth: { scopes: ['profile'] } } })
export class ProfilePage {}

// index.js
Aurelia
  .register(
    AuthGuard,
    ...,
  )
  .app(...)
  .start();
```

### 6) `canUnload`

Run an "unsaved changes" dialog guard for all components that have an `isDirty` getter on them.

- Registration scope: global

```js
// dirty-state-guard.js
@lifecycleHooks
export class DirtyStateGuard {
  constructor(@IDialogService dialog) { this.dialog = dialog; }
  async canUnload(vm, params, next, current) {
    if (vm.isDirty) {
      return this.dialog.promptUnsavedChanges();
    }
    return true;
  }
}

// index.js
Aurelia
  .register(
    DirtyStateGuard,
    ...,
  )
  .app(...)
  .start();
```

**Comment by 3cp** [comment]
Loving it.

I think `@lifecycleHooks({ order: 'before' | 'after' | 'parallel' })` could be popular enough to justify a simpler api that works with conventions too.

```js
export class SomeLIfecycleHooks {
  static order = 'before';
}
```

**Comment by jwx** [comment]
Ever since you mentioned sharing lifecycle hook code between components I've been thinking about a way to consolidate that with the hooks I've got in the router. In an attempt to explain my "merge" between this and the router hooks, I started a comment. It soon however became evident that the comment would be either confusing or huge, so instead I took your RFC here and made the changes to it. The result is https://github.com/aurelia/aurelia/issues/1047.

**Comment by fkleuver** [comment]
@jwx I've put my original markdown in a local `.md` file, committed it to git and then copied yours over it to see the diff. The differences I've found:

1. Rename `lifecycleHooks` to `hook`

In principle I'm not opposed to this. It's shorter, and by being less specific in the wording, it may be more obvious/inviting to be extended.

2. Remove the `match` function API from the "future extensions" bit and move it in-scope as an `include` / `exclude` pair,

I deliberately proposed to leave this out-of-scope for alpha because this needs more thought. We need to think not just about ergonomy here but also about type-safety, tooling integration, performance, forward compatibility with other resource types, and of course all the testing that is required when offering so many combinations of doing things.

In that regard, I prefer to err on the side of starting with too little api and expanding on as the needs/requests arise. Until then, all of the use cases are already supported by using just a little bit more code in the hooks themselves.

3. Add the `Hook.define` example with an arrow function.

FYI, you still need to take the return value and register that, so the example is a bit misleading:
```ts
const MyCanLoadHook = Hook.define({
  name: 'canLoad',
  include: { viewport: 'secured' },
  exclude: ['no-access', ContactSupport ]},
  (vm, ...args) => { /* Check auth */ });

// ...
Aurelia.register(MyCanLoadHook).app(...)
```

Normally, that return value is the class that you're decorating/defining, and then the whole DI integration is a freebie.

Support for arrow functions (as well as object literals) has been discussed on Discord a few times and been on the map for over a year now, but the current resource infrastructure is not yet attuned and tested for that, so for alpha this is not an option.

Post-alpha we'll definitely be giving some attention to less boilerplatey and more compositional mechanisms such as this, but we shouldn't have little bits and pieces flying out. This is a framework-wide concern and needs to be handled as such.

4. Add an example with a `transformTitle` hook.

Once the `transformTitle` api itself is fleshed out and properly tested & documented, of course it would be part of the game like all the other hooks, so no issues here.

**Comment by jwx** [comment]
@fkleuver It was not clear to me that this was the "alpha only" proposal. But still, the router's `HookManager`, which this in part will replace, has some of this in place and intended for alpha. We'll need to decide whether to remove the router features or include the wiring for them here.

2. Not having this forces all hooks to be either local or context aware. It would for example not be possible to do
```ts
// animations.ts
export const FadeIn = vm => { /* fade in */ };
export const FadeOut = vm => { /* fade out */ };

// auth.ts
export const Authorized = vm => { /* check basic access */ };
export const AuthorizedScope = (vm, params, instruction) => { /* check specific access from instruction.data.auth */ };

// main.ts
Aurelia.register(
  Hook.define({ exclude: 'start-page' }, FadeIn),
  Hook.define({ exclude: ['start-page', 'login', 'recover-password'] }, Authorized),
  Hook.define({ include: { viewport: 'secured' } }, AuthorizedScope),
).app(...)
```
3. All three examples in that block are regarding `include/exclude` options and registration has no bearing on that.

I envisioned the implementation of support for functions to be something like
```ts
// hook.ts
class Wrapper {
  constructor(name, method) {
    this[name] = method;
  }
  register(container) { /* ... */ }
  createContainer() { /* ... */ }
}

const Hook = {
  define(options, funcOrClass) {
    const type = isFunction(funcOrClass) ? new Wrapper(options.name, funcOrClass) : funcOrClass;
    // Do stuff with type
    return type;
};
```
This might of course not be good enough, but there should be time to find a better implementation after alpha.

4. `transformTitle` will be in place for alpha (only documentation remains).

**Comment by fkleuver** [comment]
I think my comment that it replaces v1's pipeline was wrong, because upon giving this more thought, it doesn't really work like that.

v1 pipeline slots offer specific integration points in router-global timings (not timings pertaining to specific component hooks). This proposal is about having a way to register hooks that normally only exist on components.

I was/am under the impression that HookManager was intended to be something very router-specific that would deal with hooks pertaining to specific call sites in the router, not necessarily directly tied to component hooks.

So, question then: is `transformTitle` intended to also be a component-level hook, (that would be called on component instances if it exists, like canLoad/binding/attach etc) or is it a router-level hook that's called once per navigation at a specific moment during the router's processing work?

> 2. Not having this forces all hooks to be either local or context aware. It would for example not be possible to do.

This can easily be solved by wrapping:
```ts
// animations.ts
export const FadeIn = vm => { /* fade in */ };
export const FadeOut = vm => { /* fade out */ };

// auth.ts
export const Authorized = vm => { /* check basic access */ };
export const AuthorizedScope = (vm, params, instruction) => { /* check specific access from instruction.data.auth */ };

// main.ts
Aurelia.register(
  Hook.define({ name: 'attaching' }, (vm, ...args) => {
    if (!['start-page'].includes(vm.$controller.definition.name)) {
      return FadeIn(vm, ...args);
    }
  }),
  Hook.define({ name: 'canLoad' }, (vm, params, instruction) => { // consolidated for brevity
    if (instruction.viewport === 'secured') {
      return AuthorizedScope(vm, params, instruction);
    }
    if (!['start-page', 'login', 'recover-password'].includes(instruction.path)) {
      return Authorized(vm);
    }
    return true;
  }),
).app(...)
```

> 3. All three examples in that block are regarding include/exclude options and registration has no bearing on that.

All I meant to say is that we should compare apples to apples. I included the registration bits in my examples and you didn't include it in those arrow function examples (but now you did, so that's good).

> I envisioned the implementation of support for functions to be something like

Yes, something like that, but it's something I want to leave for post-alpha because we do need to properly test stuff before we commit it to public api. I want to know for sure that it will hold up in various scenarios first, see if there are any caveats that might make the final API a bit less pretty (to avoid disappointing people)

All in all I don't disagree with these additions but I don't want to include them in scope for alpha, at least not in the `@hook` api.

**Comment by Sayan751** [comment]
Sorry. Was reading on my phone :(

**Comment by Sayan751** [comment]
Read the main RFC. Loved it. Especially the auth guard hook. 

Yet to read the discussion that followed. So, ignore if already asked. What is the specifications for dealing with multiple hooks applied to an custom element? More specifically, in which order those are executed? For example, multiple hooks with 'before' configuration.

**Comment by bigopon** [comment]
I'm curious about this part
> Unlike @viewEngineHooks, there is not a pre-determined set of hooks that can be tapped into. This allows users and plugin authors to define new hooks if they want to, in much the same way that Aurelia itself (for starters the runtime and the router) would be utilizing it.

How will it work? And will the existing resources that has lifecycle hooks (elements/attributes) will be refactored to be built on top of this?

Also, what will the equivalence of this for the template compiler?

**Comment by bigopon** [comment]
An additional use case for my comment above is the `Portal` custom attribute. It provides some extra callback bindings:
- activating
- activated
- deactivating
- deactivated

How do we use this hooks RFC in this scenario?

**Comment by jwx** [comment]
@fkleuver Since your proposal is only about component lifecycle hooks, it doesn't cover `HookManager` because `HookManager` isn't limited to lifecycle hooks or components. My proposal was (and still is) a unified hook API based on the scope and features of `HookManager` with your hook classes and registration added. Trying to clarify: my proposal isn't just your proposal with some minor changes; it's a proposal with a bigger scope that covers the features and functionality in my `HookManager` and your resource API proposal.

> So, question then: is `transformTitle` intended to also be a component-level hook, (that would be called on component instances if it exists, like canLoad/binding/attach etc)?

Setting the title involves components, viewports and the router. There'll be a router hook for `transformTitle` and probably at least for component.

> This can easily be solved by wrapping:

I think that limiting the match on globally registered hooks will be quite common and for me it's typical to the kind of things that Aurelia has a tendency to help out with. It might be an easy add for these use cases (but in my opinion already not a pretty one), but once components are allowed to be types or instances and viewports instances it gets a bit less easy (and less pretty). And, I think, common enough that it'd be a good place for Aurelia to help out.

> All I meant to say is that we should compare apples to apples.

Since these examples were specifically, and only, about the include/exclude options there's nothing in your examples to compare with. It's apples to nothing.

> All in all I don't disagree with these additions but I don't want to include them in scope for alpha, at least not in the `@hook` api.

Hm, maybe none of this is ready to be included in the scope for alpha?

---

@bigopon Within my proposal it'd be possible to consume the hooks with something like
```ts
const hookResult = Hook.invoke(type, name, instance, ...parameters);
```
where 
- `type` is the type (of instance) the hooks are tied to (for example a custom element type),
- `name` is the hook name (for example `canUnload`),
- `instance` is the relevant instance (for example view model), and 
- `parameters` are the additional, hook relevant parameters.

**Comment by bigopon** [comment]
@jwx from this example

```ts
const hookResult = Hook.invoke(type, name, instance, ...parameters);
```
How will the hook registration look like. The current usage of those callback bindings is something like this:
```html
<toast portal="target: body; activating.bind: onActivateToast; deactivating.bind: onDeactivatingToast">
```
and the view model would look like:
```ts
class MyFeed {
  onActivatingToast(target, view) {}
}
```

**Comment by jwx** [comment]
@bigopon The general idea is something like
```ts
// portal-hooks.ts (or wherever)
const portalHook = Hook.define({ type: Portal, name: 'activating' }, (target, view) => { /* hook code */ });

// my-feed.ts
class MyFeed {
  static dependencies: [portalHook];
}

// portal.ts
const hookResult = Hook.invoke(Portal, 'activating', hostTarget, hostView);
```
or
```ts
// portal-hooks.ts
@hook({ type: Portal })
class PortalHooks {
  activating(target, view) { /* hook code */ }
  deactivating(target, view) { /* hook code */ }
} 

// my-feed.ts
@customElement({ name: 'my-feed', dependencies: [PortalHooks], template })
class MyFeed { }

// portal.ts
const hookResult = Hook.invoke(Portal, 'activating', hostTarget, hostView);
const hookResult = Hook.invoke(Portal, 'deactivating', hostTarget, hostView);
```
or
```ts
// main.ts
Aurelia.register(
  Hook.define({ type: Portal, name: 'activating',  exclude: 'main' }, (target, view) => { /* hook code */ });
).app(...)

// portal.ts
const hookResult = Hook.invoke(Portal, 'activating', hostTarget, hostView);
```
I've used `Portal` as a kind of name space here so that there can be more than one hook named `activating`. Naming the hook `portalActivating` would've been an alternative to that. As seen, invoking code in portal stays the same regardless of how dev defines the hook(s).

Note that I'm not intending for the `Hook`s to replace direct bindings, callbacks or whatever fits best for the specific cases. It's  an additional way to do something from outside of the component (or whatever the target is), another way to do composition. One that the target/view model needs to do nothing at all to activate. It doesn't even need to be aware of it.

**Comment by 3cp** [comment]
I had another thought on `order: 'before' | 'after' | 'parallel'`, I think this order should be controllable per method, not per class.

For example, a plugin might provide a hook to setup and wire down some dom setup in attached and detaching, naturally it would want use "before" on attached, and "after" on detaching. (Or maybe the other way around)

**Comment by fkleuver** [comment]
@jwx 
> it's a proposal with a bigger scope that covers the features and functionality in my HookManager and your resource API proposal.

Let's say we have two concepts that we're talking about here:

-----

### Component hooks

That's what this proposal is about. Components have hooks, and this proposal allows you to define shared hook logic outside of components, and share them across components.

V1's equivalent of this is `viewEngineHooks`

### Framework hooks

That's what the runtime's `AppTask` is for: there happens to be some alignment with component hooks in terms of naming, but timing-wise they are tied to the `App` / `AppRoot` and they only run on start/stop.

V1's equivalent of this is the router pipeline slots

-----

The question is, do we truly need both?

In relation to `AppTask`, not *really*, because anything you can do with `AppTask` could also be done by registering the right type of hook in the root `App`. That's all `AppTask` really is in its current form: app-level hooks. So perhaps framework hook is not the right word for it.

We could think about an api for plugin authors to say "register this only as a dependency to the app root" and then we could get rid of `AppTask`. So, `AppTask` might very well be deprecated and removed again during alpha, once we're sure that all of its usages can be migrated to app-scoped hook registrations.

> Setting the title involves components, viewports and the router. There'll be a router hook for transformTitle and probably at least for component.

So both? Isn't the router (framework)-level hook then really just a top-level component hook? If not, how do they differ in timing and functionality?

> I think that limiting the match on globally registered hooks will be quite common and for me it's typical to the kind of things that Aurelia has a tendency to help out with

Globally registering stuff and then filtering it, is something we will be firmly recommending *against* in v2. This is an aspect of v1 that we've been working to get rid of because of how uncooperative it is with modern bundlers and with standards in general. That also goes for any kind of string-based or overly dynamic conditions (the examples I gave that have strings, are merely a courtesy for some specific questions I've received).

Or to put it differently, the recommended ways of doing things, would be doing things in such a way that makes them relatively easily statically analyzable. This is good for tooling compatibility, performance, and in general favors forward compatibility (with things of which we don't know the existence yet)

> It might be an easy add for these use cases (but in my opinion already not a pretty one)

Also not something I would recommend regardless.

The recommended way to limit the application of a shared hook would be to use local registrations. We could add it to html conventions with `<import` syntax for those who prefer that (since that's transformed by conventions to native imports anyway)

> but once components are allowed to be types or instances and viewports instances it gets a bit less easy (and less pretty).

Types or instances are ultimately quite simple to deal with by usage of star imports. It takes a bit more code, but then yields the benefit of giving you proper lazy loading, chunking, etc for free.

As for viewport instances, I don't know what to picture here. The only example you've given is related to viewport names.

> And, I think, common enough that it'd be a good place for Aurelia to help out.

I would say, common enough to have some good examples in the docs of how people can do this natively (or maybe with some common helpers) in a way that also helps out the bundlers.

> Hm, maybe none of this is ready to be included in the scope for alpha?

We need `viewEngineHooks` regardless, because people had that in v1 and will need it for migrating to v2. This proposal mainly cleans it up a bit and formalizes it a bit more, and leaves open some future extension points, but leaves the non-critical bits (that `viewEngineHooks` didn't have either) to post-alpha.

**Comment by fkleuver** [comment]
> I had another thought on `order: 'before' | 'after' | 'parallel'`, I think this order should be controllable per method, not per class.
> 
> For example, a plugin might provide a hook to setup and wire down some dom setup in attached and detaching, naturally it would want use "before" on attached, and "after" on detaching. (Or maybe the other way around)

Put the hooks in two separate classes and export them as a single record?

**Comment by bigopon** [comment]
I think this proposal lacks the compilation hook capability of view engine specifically, and the ergonomic of view engine generally. With view engine, we deal with templates & view factory as first (class) argument, and it's easy to do work needed. With this proposal, everything is behind a view model, which makes very little sense in compilation pipeline sometimes. I think if we are to proceed with this, there needs to be separate class of hooks to actually bring the old view engine capabilities back. At the very least, I think before compile needs to be there. The hook **at least** should have this basic:

```ts
interface IViewCompilerHook {
  beforeCompile(template: HTMLElement);
  // maybe add
  afterCompile(definition: CustomElementDefinition): void;
}
```

**Comment by fkleuver** [comment]
I'm omitting irrelevant arguments in the code snippets but the method signatures are identical to those of the same-named hooks on the VM itself, with the addition of the VM being the first argument.

So, the normal compile hook signatures (on the VM instance):
```ts
  define(controller, parentContainer, definition)
  hydrating(controller)
  hydrated(controller)
  created(controller)
```

Versus the signatures on classes using the lifecycle hook sharing api:
```ts
  define(vm, controller, parentContainer, definition)
  hydrating(vm, controller)
  hydrated(vm, controller)
  created(vm, controller)
```

The capabilities of these signatures are a superset of the signatures you're showing above:
```ts
controller.definition // CustomElementDefinition
controller.definition.template // HTMLElement
```

So, I disagree with your assertion that this proposal lacks the compilation hook capability. It's actually more powerful.

That said, by being more powerful it's also more complex and so I do agree that it may lack the *ergonomy*.

When I added these hooks I focused on capability first, but did not have / take the time to iterate on the ergonomy. The names and signatures can definitely be better. We should have some wiggle room with this during alpha, as we work with plugin authors and advanced users to figure out exactly what makes sense here.

This improvement should probably start with the `ICompileHooks` interface, as fixing that will indirectly "auto update" this proposal as it builds upon the existing signatures (which I think is a good thing).
